pkg/azure/validation/validation.go

```diff
@@ -0,0 +1,288 @@
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("kind"), machineSetConfig.Kind, fmt.Sprintf("must provide one of %v", allowedKinds)))
+	}
+	return allErrs
+}
+
+func validateSubnetInfo(subnetInfo api.AzureSubnetInfo, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if utils.IsEmptyString(subnetInfo.VnetName) {
+		allErrs = append(allErrs, field.Required(fldPath.Child("vnetName"), "must provide vnetName"))
+	}
+	if utils.IsEmptyString(subnetInfo.SubnetName) {
+		allErrs = append(allErrs, field.Required(fldPath.Child("subnetName"), "must provide subnetName"))
+	}
+
+	return allErrs
+}
+
+func validateProperties(properties api.AzureVirtualMachineProperties, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// validate HardwareProfile
+	allErrs = append(allErrs, validateHardwareProfile(properties.HardwareProfile, fldPath.Child("hardwareProfile"))...)
+	// validate StorageProfile
+	allErrs = append(allErrs, validateStorageProfile(properties.StorageProfile, fldPath.Child("storageProfile"))...)
+	// validate OSProfile
+	allErrs = append(allErrs, validateOSProfile(properties.OsProfile, fldPath.Child("osProfile"))...)
+	// validate availability set and vmss
+	allErrs = append(allErrs, validateAvailabilityAndScalingConfig(properties, fldPath)...)
+	return allErrs
+}
+
+func validateHardwareProfile(hwProfile api.AzureHardwareProfile, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if utils.IsEmptyString(hwProfile.VMSize) {
+		allErrs = append(allErrs, field.Required(fldPath.Child("vmSize"), "must provide vmSize"))
+	}
+	return allErrs
+}
+
+func validateStorageProfile(storageProfile api.AzureStorageProfile, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateStorageImageRef(storageProfile.ImageReference, fldPath.Child("imageReference"))...)
+	allErrs = append(allErrs, validateOSDisk(storageProfile.OsDisk, fldPath.Child("osDisk"))...)
+	allErrs = append(allErrs, validateDataDisks(storageProfile.DataDisks, fldPath.Child("dataDisks"))...)
+	return allErrs
+}
+
+func validateOSProfile(osProfile api.AzureOSProfile, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if utils.IsEmptyString(osProfile.AdminUsername) {
+		allErrs = append(allErrs, field.Required(fldPath.Child("adminUsername"), "adminUsername must be provided"))
+	}
+	return allErrs
+}
+
+func validateStorageImageRef(imageRef api.AzureImageReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	urnIsSet := !utils.IsNilOrEmptyStringPtr(imageRef.URN)
+	communityGalleryImageIDIsSet := !utils.IsNilOrEmptyStringPtr(imageRef.CommunityGalleryImageID)
+	idIsSet := !utils.IsEmptyString(imageRef.ID)
+	sharedGalleryImageIDIsSet := !utils.IsNilOrEmptyStringPtr(imageRef.SharedGalleryImageID)
+
+	atMostOnceIdentifierSet := exactlyOneShouldBeTrue(urnIsSet, communityGalleryImageIDIsSet, idIsSet, sharedGalleryImageIDIsSet)
```

Comment:
why not have the name as suggested?
```suggestion
	exactlyOneIdentifierSet := exactlyOneShouldBeTrue(urnIsSet, communityGalleryImageIDIsSet, idIsSet, sharedGalleryImageIDIsSet)
```

Reply:
Yeah previously this method was called `atMostOneShouldBeTrue` but then it was pointed out that this is not semantically correct. So i had changed the name of the method but forgot to change the variable.